url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageARMv5.java

```diff
@@ -0,0 +1,4246 @@
+        case 2:
+            //レジスタ算術右シフト
+            return String.format("%s, asr %s",
+                    getRegName(rm), getRegName(rs));
+        case 3:
+            //レジスタ右ローテート
+            return String.format("%s, ror %s",
+                    getRegName(rm), getRegName(rs));
+        default:
+            //do nothing
+            break;
+        }
+
+        throw new IllegalArgumentException("Unknown Reg Shift " +
+                String.format("0x%08x, shift:%d.",
+                        inst.getInst(), shift));
+    }
+
+    /**
+     * アドレシングモード 2 - ワードまたは符号無しバイトロード/ストア、
+     * を取得します。
+     *
+     * I ビットの意味がデータ処理命令と逆で、
+     * I=0 のときイミディエートオフセット、
+     * I=1 のときレジスタオフセットを表します。
+     *
+     * オフセット、プリインデクスの場合、
+     * アクセス先のアドレスを返します。
+     *
+     * ポストインデクスの場合、
+     * 更新後のベースレジスタの値を返します。
+     *
+     * @param inst ARM 命令
+     * @return アドレス
+     */
+    public int getAddrMode2(InstructionARM inst) {
+        boolean i = inst.getIBit();
+        boolean u = inst.getBit(23);
+        int rn = inst.getRnField();
+        int shift_imm = inst.getField(7, 5);
+        int shift = inst.getField(5, 2);
+        int offset;
+
+        if (!i) {
+            //12bits イミディエートオフセット
+            //I ビットの意味がデータ処理命令と逆なので注意！
+            offset = getAddrMode2Imm(inst);
+        } else if (shift_imm == 0 && shift == 0) {
+            //レジスタオフセット/インデクス
+            offset = getAddrMode2Reg(inst);
+        } else {
+            //スケーリング済みレジスタオフセット/インデクス
+            offset = getAddrMode2Scaled(inst);
+        }
+
+        if (!u) {
+            offset = -offset;
+        }
+
+        return getReg(rn) + offset;
+    }
+
+    /**
+     * アドレシングモード 2 - ワードまたは符号無しバイトロード/ストア、
+     * 12bits イミディエートオフセットアドレスを取得します。
+     *
+     * @param inst ARM 命令
+     * @return イミディエートオフセットアドレス
+     */
+    public int getAddrMode2Imm(InstructionARM inst) {
+        int offset12 = inst.getField(0, 12);
+
+        return offset12;
+    }
+
+    /**
+     * アドレシングモード 2 - ワードまたは符号無しバイトロード/ストア、
+     * レジスタオフセットアドレスを取得します。
+     *
+     * アドレシングモード 1 - イミディエートシフトの、
+     * shifter_operand の取得と同じ処理です。
+     *
+     * @param inst ARM 命令
+     * @return レジスタオフセットアドレス
+     */
+    public int getAddrMode2Reg(InstructionARM inst) {
+        return getAddrMode1ImmShift(inst);
+    }
+
+    /**
+     * アドレシングモード 2 - ワードまたは符号無しバイトロード/ストア、
+     * スケーリング済みレジスタオフセットアドレスを取得します。
+     *
+     * アドレシングモード 1 - イミディエートシフトの、
+     * shifter_operand の取得と同じ処理です。
+     *
+     * @param inst ARM 命令
+     * @return スケーリング済みレジスタオフセットアドレス
+     */
+    public int getAddrMode2Scaled(InstructionARM inst) {
+        return getAddrMode1ImmShift(inst);
+    }
+
+    /**
+     * アドレシングモード 2 - ワードまたは符号無しバイトロード/ストア、
+     * の文字列表記を取得します。
+     *
+     * I ビットの意味がデータ処理命令と逆で、
+     * I=0 のときイミディエートオフセット、
+     * I=1 のときレジスタオフセットを表します。
+     *
+     * @param inst ARM 命令
+     * @return アドレスの文字列表記
+     */
+    public String getAddrMode2Name(InstructionARM inst) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 769-787 into a separate method.